File: apps/android/app/src/main/java/tech/stygian/presently/oauth/AccountSearchClient.kt
package tech.stygian.presently.oauth

import java.net.URLEncoder
import org.json.JSONObject

data class AccountSuggestion(
    val did: String,
    val handle: String,
    val displayName: String?,
    val avatarUrl: String?,
)

class AccountSearchClient(
    private val http: OAuthHttpClient = OAuthHttpClient(),
) {
    suspend fun search(query: String, limit: Int = 6): List<AccountSuggestion> {
        val normalized = query.trim().removePrefix("@")
        if (normalized.length < 2) return emptyList()
        val encoded = URLEncoder.encode(normalized, Charsets.UTF_8.name())
        val response = http.get(
            "$Endpoint?q=$encoded&limit=${limit.coerceIn(1, 10)}",
            mapOf("Accept" to "application/json"),
        )
        if (response.statusCode != 200) return emptyList()

        return parseResponse(String(response.body, Charsets.UTF_8))
    }

    companion object {
        const val Endpoint =
            "https://public.api.bsky.app/xrpc/app.bsky.actor.searchActorsTypeahead"

        internal fun parseResponse(body: String): List<AccountSuggestion> {
            val actors = JSONObject(body)
                .optJSONArray("actors") ?: return emptyList()
            return buildList {
                for (index in 0 until actors.length()) {
                    val actor = actors.optJSONObject(index) ?: continue
                    val did = actor.optString("did")
                    val handle = actor.optString("handle")
                    if (did.isBlank() || handle.isBlank()) continue
                    add(
                        AccountSuggestion(
                            did = did,
                            handle = handle,
                            displayName = actor.optionalString("displayName"),
                            avatarUrl = actor.optionalString("avatar"),
                        ),
                    )
                }
            }
        }
    }
}

File: apps/ios/Presently/OAuth/AccountSearchClient.swift
import Foundation

struct AccountSuggestion: Decodable, Equatable, Identifiable, Sendable {
    let did: String
    let handle: String
    let displayName: String?
    let avatar: URL?

    var id: String { did }
}

struct AccountSuggestionsResponse: Decodable, Equatable, Sendable {
    let actors: [AccountSuggestion]
}

enum AccountSearchError: Error {
    case invalidURL
}

struct AccountSearchClient: Sendable {
    private static let endpoint = URL(
        string: "https://public.api.bsky.app/xrpc/app.bsky.actor.searchActorsTypeahead"
    )!

    private let http: OAuthHTTPClient

    init(http: OAuthHTTPClient = OAuthHTTPClient()) {
        self.http = http
    }

    func suggestions(for query: String, limit: Int = 6) async throws
        -> [AccountSuggestion] {
        let response: AccountSuggestionsResponse = try await http.getJSON(
            Self.requestURL(for: query, limit: limit)
        )
        return response.actors
    }

    static func requestURL(for query: String, limit: Int = 6) throws -> URL {
        guard var components = URLComponents(
            url: endpoint,
            resolvingAgainstBaseURL: false
        ) else {
            throw AccountSearchError.invalidURL
        }
        components.queryItems = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "limit", value: String(limit)),
        ]
        guard let url = components.url else {
            throw AccountSearchError.invalidURL
        }
        return url
    }
}
